File: app/src/main/java/com/nirbhor/app/ui/screens/meds/CabinetScreen.kt
package com.nirbhor.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.statusBars
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.nirbhor.app.domain.Medicine
import com.nirbhor.app.domain.StockStatus
import com.nirbhor.app.domain.TimeBlock
import com.nirbhor.app.navigation.LocalAppContainer
import com.nirbhor.app.navigation.NavActions
import com.nirbhor.app.ui.components.NbCard
import com.nirbhor.app.ui.components.PrimaryButton
import com.nirbhor.app.ui.components.SecondaryButton
import com.nirbhor.app.ui.components.StatusPill
import com.nirbhor.app.ui.i18n.num
import com.nirbhor.app.ui.i18n.tr
import com.nirbhor.app.ui.i18n.LocalIsBangla
import com.nirbhor.app.ui.marks.MedicineMark
import com.nirbhor.app.ui.theme.Dimens
import com.nirbhor.app.ui.theme.NirbhorTheme

/** Medicine cabinet (2s/2e). Main tab — the bottom nav is provided by the app root. */
@Composable
fun CabinetScreen(actions: NavActions) {
    val colors = NirbhorTheme.colors
    val repo = LocalAppContainer.current.repository
    val medicines by repo.medicines.collectAsStateWithLifecycle(initialValue = emptyList())
    val stock by repo.stockStatuses().collectAsStateWithLifecycle(initialValue = emptyList())
    val stockById = remember(stock) { stock.associateBy { it.medicineId } }

    Column(Modifier.fillMaxSize().background(colors.paper)) {
        Row(
            Modifier
                .fillMaxWidth().background(colors.paper)
                .windowInsetsPadding(WindowInsets.statusBars)
                .padding(horizontal = Dimens.screenPadding, vertical = 16.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(tr("ওষুধের তালিকা", "Medicines"), style = NirbhorTheme.type.titleHero, color = colors.ink)
            Spacer(Modifier.width(10.dp))
            Box(Modifier.clip(RoundedCornerShape(9.dp)).background(colors.sage).padding(horizontal = 10.dp, vertical = 4.dp)) {
                Text(num(medicines.size), style = NirbhorTheme.type.statusPill, color = colors.ink2)
            }
        }

        LazyColumn(
            Modifier.weight(1f).fillMaxWidth().padding(horizontal = Dimens.screenPadding),
            verticalArrangement = Arrangement.spacedBy(Dimens.cardGap),
        ) {
            items(medicines, key = { it.id }) { med ->
                MedicineRow(med, stockById[med.id]) { actions.openMedicine(med.id) }
            }
            item { Spacer(Modifier.size(8.dp)) }
        }

        Column(
            Modifier.fillMaxWidth().padding(horizontal = Dimens.screenPadding).padding(bottom = 16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            PrimaryButton(tr("ওষুধ যোগ করুন", "Add a medicine"), actions::startAddMedicine, height = 60.dp, modifier = Modifier.fillMaxWidth())
            SecondaryButton(tr("পাতা বা বাক্স স্ক্যান করুন", "Scan a pack or box"), actions::startScan, height = 60.dp, modifier = Modifier.fillMaxWidth())
        }
    }
}

@Composable
private fun MedicineRow(med: Medicine, stock: StockStatus?, onClick: () -> Unit) {
    val colors = NirbhorTheme.colors
    NbCard(onClick = onClick) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Box(
                Modifier.size(44.dp).clip(RoundedCornerShape(12.dp)).background(colors.calmSoft),
                contentAlignment = Alignment.Center,
            ) {
                MedicineMark(shape = med.mark, color = Color(med.markColor), size = 27.dp)
            }
            Spacer(Modifier.width(12.dp))
            Column(Modifier.weight(1f)) {
                Text(med.displayName, style = NirbhorTheme.type.cardTitleSecondary, color = colors.ink)
                Text(scheduleSummary(med), style = NirbhorTheme.type.meta, color = colors.ink3)
            }
            if (stock != null && stock.isLow) {
                StatusPill(tr("${num(stock.count)}টি বাকি", "${stock.count} left"), colors.warmSoft, colors.warmD)
            } else {
                Icon(
                    Icons.AutoMirrored.Filled.KeyboardArrowRight,
                    contentDescription = tr("ওষুধের বিস্তারিত খুলুন", "Open medicine details"),
                    tint = colors.ink3,
                    modifier = Modifier.size(24.dp),
                )
            }
        }
    }
}

@Composable
private fun scheduleSummary(med: Medicine): String {
    // Words, not clock times — e.g. "প্রতিদিন সকাল, দুপুর, রাত".
    val bangla = LocalIsBangla.current
    val blocks = med.timeTokens.map { TimeBlock.fromToken(it) }
    val frequency = when (med.frequency) {
        com.nirbhor.app.domain.Frequency.DAILY -> if (bangla) "প্রতিদিন" else "Daily"
        com.nirbhor.app.domain.Frequency.ALTERNATE -> if (bangla) "একদিন পরপর" else "Every other day"
        com.nirbhor.app.domain.Frequency.WEEKDAYS -> if (bangla) "সপ্তাহের কর্মদিবসে" else "On weekdays"
        com.nirbhor.app.domain.Frequency.WEEKLY -> if (bangla) "নির্বাচিত দিনে" else "On selected days"
    }
    val times = blocks.joinToString(", ") {
        when (it) {
            TimeBlock.MORNING -> if (bangla) "সকাল" else "morning"
            TimeBlock.NOON -> if (bangla) "দুপুর" else "afternoon"
            TimeBlock.NIGHT -> if (bangla) "রাত" else "night"
        }
    }
    return "$frequency · $times"
}

File: lib/ui/screens/cabinet_screen.dart
import 'package:flutter/material.dart';

import '../../data/app_scope.dart';
import '../../domain/models.dart';
import '../../navigation/nav_actions.dart';
import '../../theme/theme.dart';
import '../components/buttons.dart';
import '../components/labels.dart';
import '../components/surfaces.dart';
import '../marks/medicine_mark.dart';

/// Medicine cabinet (2s/2e) — every medicine, its schedule in words, and its stock warning.
class CabinetScreen extends StatelessWidget {
  const CabinetScreen({super.key});

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    return ColoredBox(
      color: colors.paper,
      child: RepoBuilder<(List<Medicine>, List<StockStatus>)>(
        query: (repo) async => (await repo.medicines(), await repo.stockStatuses()),
        builder: (context, data) {
          final (medicines, stock) = data;
          final stockById = {for (final s in stock) s.medicineId: s};

          return Column(
            children: [
              SafeArea(
                bottom: false,
                child: Padding(
                  padding: const EdgeInsets.symmetric(
                    horizontal: Dimens.screenPadding,
                    vertical: 16,
                  ),
                  child: Row(
                    children: [
                      Text(
                        context.tr('ওষুধের তালিকা', 'Medicines'),
                        style: context.type.titleHero.copyWith(color: colors.ink),
                      ),
                      const SizedBox(width: 10),
                      Container(
                        decoration: BoxDecoration(
                          color: colors.sage,
                          borderRadius: BorderRadius.circular(9),
                        ),
                        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
                        child: Text(
                          context.num(medicines.length),
                          style: context.type.statusPill.copyWith(color: colors.ink2),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
              Expanded(
                child: medicines.isEmpty
                    ? Padding(
                        padding: const EdgeInsets.symmetric(horizontal: Dimens.screenPadding),
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.stretch,
                          children: [
                            const SizedBox(height: 8),
                            Text(
                              context.tr(
                                'এখনও কোনো ওষুধ যোগ করা হয়নি',
                                'No medicines added yet',
                              ),
                              style: context.type.cardTitlePrimary.copyWith(color: colors.ink),
                            ),
                            const SizedBox(height: 10),
                            Text(
                              context.tr(
                                'পাতা স্ক্যান করলে নাম আর শক্তি নিজেই পড়ে নেবে — অথবা নাম লিখে যোগ করুন।',
                                'Scanning a pack reads the name and strength for you — or add one by typing its name.',
                              ),
                              style: context.type.body.copyWith(color: colors.ink2),
                            ),
                          ],
                        ),
                      )
                    : ListView.separated(
                        padding: const EdgeInsets.fromLTRB(
                          Dimens.screenPadding,
                          0,
                          Dimens.screenPadding,
                          8,
                        ),
                        itemCount: medicines.length,
                        separatorBuilder: (_, _) => const SizedBox(height: Dimens.cardGap),
                        itemBuilder: (context, i) {
                          final med = medicines[i];
                          return _MedicineRow(
                            med: med,
                            stock: stockById[med.id],
                            onTap: () => context.nav.openMedicine(med.id),
                          );
                        },
                      ),
              ),
              Padding(
                padding: const EdgeInsets.fromLTRB(
                  Dimens.screenPadding,
                  0,
                  Dimens.screenPadding,
                  16,
                ),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  children: [
                    PrimaryButton(
                      text: context.tr('ওষুধ যোগ করুন', 'Add a medicine'),
                      onPressed: context.nav.startAddMedicine,
                      height: 60,
                    ),
                    const SizedBox(height: 8),
                    SecondaryButton(
                      text: context.tr('পাতা বা বাক্স স্ক্যান করুন', 'Scan a pack or box'),
                      onPressed: context.nav.startScan,
                      height: 60,
                    ),
                  ],
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}

class _MedicineRow extends StatelessWidget {
  const _MedicineRow({required this.med, required this.stock, required this.onTap});

  final Medicine med;
  final StockStatus? stock;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    final low = stock != null && stock!.isLow;
    return NbCard(
      onTap: onTap,
      child: Row(
        children: [
          Container(
            width: 44,
            height: 44,
            alignment: Alignment.center,
            decoration: BoxDecoration(
              color: colors.calmSoft,
              borderRadius: BorderRadius.circular(12),
            ),
            child: MedicineMark(shape: med.mark, color: Color(med.markColor), size: 27),
          ),
          const SizedBox(width: 12),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  med.displayName,
                  style: context.type.cardTitleSecondary.copyWith(color: colors.ink),
                ),
                Text(
                  scheduleSummary(context, med),
                  style: context.type.meta.copyWith(color: colors.ink3),
                ),
              ],
            ),
          ),
          if (low)
            StatusPill(
              text: context.tr(
                '${context.num(stock!.count)}টি বাকি',
                '${stock!.count} left', hi: '${stock!.count} बची', es: 'quedan ${stock!.count}',
              ),
              background: colors.warmSoft,
              contentColor: colors.warmD,
            )
          else
            Semantics(
              label: context.tr('ওষুধের বিস্তারিত খুলুন', 'Open medicine details'),
              child: Icon(Icons.keyboard_arrow_right, size: 24, color: colors.ink3),
            ),
        ],
      ),
    );
  }
}

/// Words, not clock times — e.g. "প্রতিদিন · সকাল, দুপুর, রাত".
String scheduleSummary(BuildContext context, Medicine med) {
  final frequency = switch (med.frequency) {
    Frequency.daily => context.tr('প্রতিদিন', 'Daily'),
    Frequency.alternate => context.tr('একদিন পরপর', 'Every other day'),
    Frequency.weekdays => context.tr('সপ্তাহের কর্মদিবসে', 'On weekdays'),
    Frequency.weekly => context.tr('নির্বাচিত দিনে', 'On selected days'),
  };
  final times = med.timeTokens.map((t) {
    return switch (TimeBlock.fromToken(t)) {
      TimeBlock.morning => context.tr('সকাল', 'morning'),
      TimeBlock.noon => context.tr('দুপুর', 'afternoon'),
      TimeBlock.night => context.tr('রাত', 'night'),
    };
  }).join(', ');
  return '$frequency · $times';
}
